electrum_nmc/electrum/names.py

```diff
@@ -131,6 +133,127 @@ def validate_value_length(value: bytes):
+
+        # Check if the truncated checksum matches the corresponding bytes in the decoded service ID
+        if truncated_checksum[0] != decoded_service_id[V3_ONION_SERVICE_ID_CHECKSUM_OFFSET] or \
+                truncated_checksum[1] != decoded_service_id[V3_ONION_SERVICE_ID_CHECKSUM_OFFSET + 1]:
+            raise ValueError("Invalid checksum")
+    
+    except Exception as e:
+        raise ValueError(f"Invalid onion address: {e}")
+
+def calc_truncated_checksum(public_key):
+    # Define the size of the hash in bytes
+    SHA256_BYTES = 256 // 8
+    hash_bytes = bytearray(SHA256_BYTES)
+
+    hasher = hashlib.sha3_256()
+    assert SHA256_BYTES == hasher.digest_size
+
+    # Calculate the checksum
+    hasher.update(b".onion checksum")
+    hasher.update(public_key)
+    hasher.update(bytes([0x03]))
+    hash_bytes = bytearray(hasher.digest())
+
+    return hash_bytes[:2]
+
+def validate_zeronet_address(address: str) -> None: # Validate P2PKH Address (Zero Net)
+    try:
+        addrtype, _ = b58_address_to_hash160(address)
+    except Exception as e:
+        raise ValueError(f"Invalid zeronet address: {e}")
+    
+    if addrtype != 0:
+        raise ValueError("Invalid address type.")
+
+def validate_ip_address(address: str) -> None:
+    try:
+        ipaddress.ip_address(address)
+    except Exception as e:
+        raise ValueError(f"Invalid IP address: {e}")
+
+def validate_i2p_address(address: str) -> None: # Naming rules were implemeted from https://geti2p.net/en/docs/naming
+    try:
+        # Check length
+        if len(address) > 67:
+            raise ValueError("Address length exceeds 67 characters")
+
+        # Check if it ends with '.b32.i2p'
+        if address.endswith('.b32.i2p'):
+            validate_base32(address)
+            return  # End function if base32 validation passes
+        
+        # Check if it ends with '.i2p' (after base32 check)
+        if not address.endswith('.i2p'):
+            raise ValueError("Address must end with '.i2p'")
+
+
+        # Check if it starts with '.' or '-'
+        if address.startswith('.') or address.startswith('-'):
+            raise ValueError("Address cannot start with '.' or '-'")
+
+        # Check for consecutive dots, '.-' or '-.'
+        if '..' in address or '.-' in address or '-.' in address:
+            raise ValueError("Address contains invalid sequences")
+
+        # Check for valid characters using regex. Allows only small letters, number, '.' and '-'
+        if not re.match(r'^[a-z0-9.\-]+$', address):
+            raise ValueError("Address contains invalid characters.")
+
+        # Check for IDN format. '--' should only be present in 'xn--' 
+        if '--' in address and ('xn--' not in address or '--' in address[4:]):
+            raise ValueError("Invalid IDN format")
+    except ValueError as e:
+        raise ValueError(f"Invalid I2P Address: {e}")
+
+def validate_base32(address: str) -> None:
```

Comment:
I think we can merge this function with `validate_i2p_address`. Also, the magic constant `-8` is not great for readability, maybe replace with `len('.b32.i2p')`

Reply:
I2P's docs say:

> I2P uses 52 characters (256 bits) to represent the full SHA-256 hash. The form is {52 chars}.b32.i2p.

I'm not sure what drugs the I2P devs were on when they wrote that, seeing as a base32 character is 5 bits, and 52*5 is 260 bits, not 256. I assume the remaining bits are used as a checksum like in Tor. If so, we should presumably verify that checksum.